posts/01-effective-rust-canisters.html.pm

```diff
@@ -442,6 +442,355 @@ type CreateEntityResult = variant {
 }
 }
 
+◊section{
+◊section-title["optimization"]{Optimization}
+◊subsection-title["cycle-consumption"]{Reducing cycle consumption}
+◊p{
+  The first step towards an optimized system is profiling.
+}
+
+◊advice["performance-counter"]{Measure the number of instructions your endpoints consume.}
+
+◊p{
+  The ◊a[#:href "https://docs.rs/ic-cdk/0.5.2/ic_cdk/api/call/fn.performance_counter.html"]{◊code{performance_counter}} API will tell you the number of ◊em{instructions} your code consumed since the last ◊a[#:href "https://internetcomputer.org/docs/current/references/ic-interface-spec/#entry-points"]{entry point}.
+  Instructions are the internal currency of the IC runtime: the system defines all its limits in terms of instructions.
+  As of July 2022, the limits are:
+}
+◊ul[#:class "arrows"]{
+◊li{One message execution: ◊a[#:href "https://github.com/dfinity/ic/blob/7d3fb4ef01416241205818450156aabd21c24b34/rs/config/src/subnet_config.rs#L19"]{5 billion} instructions.}
+◊li{One round◊sidenote["sn-round"]{Each block produced by consensus initiates a round of execution.}: ◊a[#:href "https://github.com/dfinity/ic/blob/7d3fb4ef01416241205818450156aabd21c24b34/rs/config/src/subnet_config.rs#L46"]{7 billion} instructions.}
+◊li{Canister upgrade: ◊a[#:href "https://github.com/dfinity/ic/blob/7d3fb4ef01416241205818450156aabd21c24b34/rs/config/src/subnet_config.rs#L56"]{200 billion} instructions.}
+}
+◊p{
+  Instructions are not cycles, but there is a ◊a[#:href "https://github.com/dfinity/ic/blob/c01d7d1b2e18490a2f70d2fdf5b6aceccab5860c/rs/cycles_account_manager/src/lib.rs#L730-L738"]{simple linear function} that converts instructions to cycles.
```

Comment:
It will change soon, with DTS we plan to charge more for long executions. I'm not sure we have any public document or code for that though...

Reply:
It would be great to have an official document to which I can refer. I wrote down these limits because they are handy to know but not documented anywhere.